lms/djangoapps/courseware/tests/test_module_render.py

```diff
@@ -2637,6 +2638,40 @@ def test_anonymous_student_id(self):
         )
         assert runtime.anonymous_student_id == anonymous_id_for_user(self.user, self.course.id)
 
+    def test_anonymous_student_id_bug(self):
+        """
+        Verifies that subsequent calls to get_module_system_for_user have no effect on each block runtime's
+        anonymous_student_id value.
+        """
+        problem_runtime, _ = render.get_module_system_for_user(
+            self.user,
+            self.student_data,
+            self.problem_descriptor,
+            self.course.id,
+            self.track_function,
+            self.xqueue_callback_url_prefix,
+            self.request_token,
+            course=self.course,
+        )
+        # Ensure the problem block returns a per-user anonymous id
+        assert problem_runtime.anonymous_student_id == anonymous_id_for_user(self.user, None)
+
+        vertical_runtime, _ = render.get_module_system_for_user(
+            self.user,
+            self.student_data,
+            self.descriptor,
+            self.course.id,
+            self.track_function,
+            self.xqueue_callback_url_prefix,
+            self.request_token,
+            course=self.course,
+        )
+        # Ensure the vertical block returns a per-course+user anonymous id
+        assert vertical_runtime.anonymous_student_id == anonymous_id_for_user(self.user, self.course.id)
+
+        # Ensure the problem runtime's anonymous student ID is unchanged after the above call.
+        assert problem_runtime.anonymous_student_id == anonymous_id_for_user(self.user, None)
```

Comment:
Without a14f7e59531108039d32d6f760d8b6197a0c9dc1. this test fails with:
```
AssertionError: assert 'e9018d2873e89d4dc26ea4d7ef631a1f' == '8538ca55520679e6d31b0c709d19e640'
  - 8538ca55520679e6d31b0c709d19e640
  + e9018d2873e89d4dc26ea4d7ef631a1f

Stacktrace

self = <lms.djangoapps.courseware.tests.test_module_render.LmsModuleSystemShimTest testMethod=test_anonymous_student_id_bug>

    def test_anonymous_student_id_bug(self):
        """
        Verifies that subsequent calls to get_module_system_for_user have no effect on each block runtime's
        anonymous_student_id value.
        """
        problem_runtime, _ = render.get_module_system_for_user(
            self.user,
            self.student_data,
            self.problem_descriptor,
            self.course.id,
            self.track_function,
            self.xqueue_callback_url_prefix,
            self.request_token,
            course=self.course,
        )
        # Ensure the problem block returns a per-user anonymous id
        assert problem_runtime.anonymous_student_id == anonymous_id_for_user(self.user, None)

        vertical_runtime, _ = render.get_module_system_for_user(
            self.user,
            self.student_data,
            self.descriptor,
            self.course.id,
            self.track_function,
            self.xqueue_callback_url_prefix,
            self.request_token,
            course=self.course,
        )
        # Ensure the vertical block returns a per-course+user anonymous id
        assert vertical_runtime.anonymous_student_id == anonymous_id_for_user(self.user, self.course.id)

        # Ensure the problem runtime's anonymous student ID is unchanged after the above call.
>       assert problem_runtime.anonymous_student_id == anonymous_id_for_user(self.user, None)
E       AssertionError: assert 'e9018d2873e89d4dc26ea4d7ef631a1f' == '8538ca55520679e6d31b0c709d19e640'
E         - 8538ca55520679e6d31b0c709d19e640
E         + e9018d2873e89d4dc26ea4d7ef631a1f

lms/djangoapps/courseware/tests/test_module_render.py:2673: AssertionError
```

See https://build.testeng.edx.org/job/edx-platform-python-pipeline-pr/36211/testReport/junit/lms.djangoapps.courseware.tests.test_module_render/LmsModuleSystemShimTest/Run_Tests___lms_unit___test_anonymous_student_id_bug/